### ISSUE TITLE
failed to install on Windows

### ISSUE THREAD

**Issue by altavir** [twosigma/beakerx#5075]
Using `gradle install` in `gradlex` gives a error.
I believe that the main problem is: `npm unavailable.  If you're running this command using sudo, make sure npm is available to sudo`
Npm is installed and accessible via `npm` command from the same console.

**Comment by scottdraves** [comment]
sorry not sure what you mean.  we don't have an install target, and what is gradlex?

**Comment by altavir** [comment]
Sorry, was writing in a hurry. I mean `beakerx` and it has `install` task:
```
task install() {
  doLast {
    exec {
      commandLine(*(winOrUnix('pip', "install", "-e", ".")))
    }
    exec {
      commandLine(*(winOrUnix("jupyter", "nbextension", "install", "--py", "--symlink", "--sys-prefix", "beakerx")))
    }
    exec {
      commandLine(*(winOrUnix("jupyter", "nbextension", "enable", "--py", "--sys-prefix", "beakerx")))
    }
    project('js').tasks.install.execute()
  }
}
```

**Comment by altavir** [comment]
The same result if I use `linkBeakerNbextension` from root project

**Comment by scottdraves** [comment]
can you provide a complete log?
it should work.

**Comment by altavir** [comment]
Here is the log:
```
:beakerx:install
Obtaining file:///D:/beakerx/beakerx
    Complete output from command python setup.py egg_info:
    setup.py entered
    $PATH=C:\Anaconda3\Library\bin;C:\Anaconda3\Library\bin;C:\Program Files\ConEmu\ConEmu\Scripts;C:\Program Files\ConEmu;C:\Program Files\ConEmu\ConEmu;C:\ProgramData\Oracle\Java\javapath;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\Program Files\TortoiseHg\;C:\Anaconda3;C:\Anaconda3\Scripts;C:\Anaconda3\Library\bin;C:\Program Files (x86)\Skype\Phone\;C:\Program Files\Calibre2\;C:\Program Files\jEdit;C:\Program Files (x86)\Pandoc\;C:\Program Files\Git\cmd;C:\Program Files (x86)\PuTTY\;C:\Program Files\nodejs\;C:\Users\darksnake\AppData\Local\atom\bin;C:\Users\darksnake\AppData\Local\Microsoft\WindowsApps;C:\kotlinc\bin;C:\Users\darksnake\.posh_gvm\gradle\current\bin;C:\Users\darksnake\AppData\Roaming\npm
    running egg_info
    running jsdeps
    C:\Anaconda3\lib\site-packages\setuptools-20.3-py3.5.egg\setuptools\dist.py:285: UserWarning: Normalizing '0.1.0.dev' to '0.1.0.dev0'
    `npm` unavailable.  If you're running this command using sudo, make sure `npm` is available to sudo
    rebuilding js and css failed
    missing files: ['D:\\beakerx\\beakerx\\beakerx\\static\\extension.js', 'D:\\beakerx\\beakerx\\beakerx\\static\\index.js']
Command "python setup.py egg_info" failed with error code 1 in D:\beakerx\beakerx\    Traceback (most recent call last):

      File "<string>", line 1, in <module>
      File "D:\beakerx\beakerx\setup.py", line 166, in <module>
        setup(**setup_args)
      File "C:\Anaconda3\lib\distutils\core.py", line 148, in setup
        dist.run_commands()
      File "C:\Anaconda3\lib\distutils\dist.py", line 955, in run_commands
        self.run_command(cmd)
      File "C:\Anaconda3\lib\distutils\dist.py", line 974, in run_command
        cmd_obj.run()
      File "D:\beakerx\beakerx\setup.py", line 45, in run
        raise e
      File "D:\beakerx\beakerx\setup.py", line 38, in run
        self.distribution.run_command('jsdeps')
      File "C:\Anaconda3\lib\distutils\dist.py", line 974, in run_command
        cmd_obj.run()
      File "D:\beakerx\beakerx\setup.py", line 109, in run
        raise ValueError(msg)
    ValueError: Missing file: D:\beakerx\beakerx\beakerx\static\extension.js
    npm is required to build a development version of widgetsnbextension

    ----------------------------------------
You are using pip version 8.1.1, however version 9.0.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.
:beakerx:install FAILED

FAILURE: Build failed with an exception.

* Where:
Build file 'D:\beakerx\beakerx\build.gradle' line: 31

* What went wrong:
Execution failed for task ':beakerx:install'.
> Process 'command 'cmd'' finished with non-zero exit value 1
```

It is form different computer with the same setup (win 10, Anaconda 3, npm). The result is the same.

**Comment by scottdraves** [comment]
please provide the log for what happens if you run `gradle linkBeakerNbextension`

**Comment by altavir** [comment]
Exactly the same:

```
:linkBeakerNbextension
Obtaining file:///D:/Work/beakerx/beakerx
    Complete output from command python setup.py egg_info:
    setup.py entered
Command "python setup.py egg_info" failed with error code 1 in D:\Work\beakerx\beakerx\
    $PATH=C:\Anaconda3\Library\bin;C:\Anaconda3\Library\bin;C:\Program Files\ConEmu\ConEmu\Scripts;C:\ProgramData\Oracle\Java\javapath;C:\Program Files\ConEmu;C:\Program Files\ConEmu\ConEmu;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;c:\Program Files\Java\jdk\\bin;C:\Program Files (x86)\AMD APP\bin\x86_64;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Program Files (x86)\ATI Technologies\ATI.ACE\Core-Static;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit\;C:\Program Files\Calibre2\;C:\Program Files\Mercurial\;C:\Program Files\Git\cmd;C:\Users\darksnake\.posh_gvm\gradle\current\bin;C:\Users\darksnake\.posh_gvm\grails\current\bin;C:\Users\darksnake\.posh_gvm\groovy\current\bin;C:\MinGW\bin;C:\Users\darksnake\AppData\Local\atom\bin;C:\Users\darksnake\AppData\Roaming\npm;C:\Program Files\Docker Toolbox;C:\Program Files\ImageMagick-7.0.0-Q16;C:\Program Files\jEdit;C:\Program Files\nodejs\;C:\Users\darksnake\AppData\Local\Microsoft\WindowsApps;C:\Program Files (x86)\Pandoc\;C:\Program Files\MiKTeX 2.9\miktex\bin\x64\;C:\Program Files\TortoiseHg\;C:\Program Files (x86)\Skype\Phone\;C:\Anaconda3;C:\Anaconda3\Scripts;C:\Anaconda3\Library\bin;C:\Users\darksnake\.posh_gvm\gradle\current\bin;C:\Users\darksnake\.posh_gvm\grails\current\bin;C:\Users\darksnake\.posh_gvm\groovy\current\bin;C:\MinGW\bin;C:\maven\bin;C:\Users\darksnake\.posh_gvm\jbake\current\bin;C:\Users\darksnake\AppData\Local\atom\bin;C:\Users\darksnake\AppData\Roaming\npm;C:\Program Files\ImageMagick-7.0.0-Q16;C:\Program Files\jEdit;C:\Program Files\Calibre2\;C:\Program Files\nodejs\;C:\Users\darksnake\AppData\Local\Microsoft\WindowsApps;C:\Program Files\CMake\bin
    running egg_info
    running jsdeps
    C:\Anaconda3\lib\site-packages\setuptools-27.2.0-py3.6.egg\setuptools\dist.py:331: UserWarning: Normalizing '0.1.0.dev' to '0.1.0.dev0'
    `npm` unavailable.  If you're running this command using sudo, make sure `npm` is available to sudo
    rebuilding js and css failed
    missing files: ['D:\\Work\\beakerx\\beakerx\\beakerx\\static\\extension.js', 'D:\\Work\\beakerx\\beakerx\\beakerx\\static\\index.js']
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "D:\Work\beakerx\beakerx\setup.py", line 166, in <module>
        setup(**setup_args)
      File "C:\Anaconda3\lib\distutils\core.py", line 148, in setup
        dist.run_commands()
      File "C:\Anaconda3\lib\distutils\dist.py", line 955, in run_commands
        self.run_command(cmd)
      File "C:\Anaconda3\lib\distutils\dist.py", line 974, in run_command
        cmd_obj.run()
      File "D:\Work\beakerx\beakerx\setup.py", line 45, in run
        raise e
      File "D:\Work\beakerx\beakerx\setup.py", line 38, in run
        self.distribution.run_command('jsdeps')
      File "C:\Anaconda3\lib\distutils\dist.py", line 974, in run_command
        cmd_obj.run()
      File "D:\Work\beakerx\beakerx\setup.py", line 109, in run
        raise ValueError(msg)
    ValueError: Missing file: D:\Work\beakerx\beakerx\beakerx\static\extension.js
    npm is required to build a development version of widgetsnbextension

    ----------------------------------------
:linkBeakerNbextension FAILED

FAILURE: Build failed with an exception.

* Where:
Build file 'D:\Work\beakerx\beakerx\build.gradle' line: 31

* What went wrong:
Execution failed for task ':beakerx:install'.
> Process 'command 'cmd'' finished with non-zero exit value 1
```

According to `build.gradle`, it is just an alias to `:beakerx:install`, so it could not be the other way around.

**Comment by altavir** [comment]
By the way `:kernel:groovy:install` works fine. It installs groovy console which actually works. Beaker features report errors though.

**Comment by altavir** [comment]
OK, I localized the problem: `check_call(['npm', '--version'])` produces an error even if I am running it from the IPython console. `npm --version` from system terminal works fine.

**Comment by altavir** [comment]
Fixed the problem by replacing `check_call(['npm', '--version'])` with `check_call(['npm', '--version'], shell = True)`. Now I get the following:

```
PS D:\beakerx> gradle linkBeakerNbextension
:linkBeakerNbextension
Obtaining file:///D:/beakerx/beakerx
    Complete output from command python setup.py egg_info:
    setup.py entered
    $PATH=C:\Anaconda3\Library\bin;C:\Anaconda3\Library\bin;C:\Program Files\ConEmu\ConEmu\Scripts;C:\Program Files\ConEmu;C:\Program Files\ConEmu\ConEmu;C:\ProgramData\Oracle\Java\javapath;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\Program Files\TortoiseHg\;C:\Program Files (x86)\Skype\Phone\;C:\Program Files\Calibre2\;C:\Program Files\jEdit;C:\Program Files (x86)\Pandoc\;C:\Program Files\Git\cmd;C:\Program Files (x86)\PuTTY\;C:\Program Files\nodejs\;C:\Anaconda3;C:\Anaconda3\Scripts;C:\Anaconda3\Library\bin;C:\Users\darksnake\AppData\Local\atom\bin;C:\Users\darksnake\AppData\Local\Microsoft\WindowsApps;C:\Users\darksnake\.posh_gvm\gradle\current\bin;C:\Users\darksnake\AppData\Roaming\npm
    running egg_info
    running jsdeps
    3.10.10
    3.10.10
    Installing build dependencies with npm.  This may take a while...
    C:\Anaconda3\lib\site-packages\setuptools-27.2.0-py3.6.egg\setuptools\dist.py:331: UserWarning: Normalizing '0.1.0.dev' to '0.1.0.dev0'
    rebuilding js and css failed
    missing files: ['D:\\beakerx\\beakerx\\beakerx\\static\\extension.js', 'D:\\beakerx\\beakerx\\beakerx\\static\\index.js']
    error: [WinError 2] The system cannot find the file specified

    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in D:\beakerx\beakerx\
:linkBeakerNbextension FAILED

FAILURE: Build failed with an exception.
```

**Comment by altavir** [comment]
Still not working:

```
npm ERR! Windows_NT 10.0.15063
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "i"
npm ERR! node v4.4.4
npm ERR! npm  v2.15.1
npm ERR! code EPEERINVALID

npm ERR! peerinvalid The package webpack@1.15.0 does not satisfy its siblings' peerDependencies requirements!
npm ERR! peerinvalid Peer watch-ignore-webpack-plugin@1.0.0 wants webpack@1.5.x
npm ERR! peerinvalid Peer bower-webpack-plugin@0.1.9 wants webpack@>=1.4.0
npm ERR! peerinvalid Peer sass-loader@6.0.3 wants webpack@^2.0.0

npm ERR! Please include the following file with any support request:
npm ERR!     D:\Work\beakerx\beakerx\js\npm-debug.log
:beakerx:install FAILED
```

**Comment by scottdraves** [comment]
thanks for the report -- this is quite baffling, can anyone reproduce?
did you try a fresh clone?

**Comment by altavir** [comment]
Not working even with clean clone (which takes a quite a lot of time to download and build due to #5074  and npm calls). 
By the way, the problem with `check_call` is still not solved. Without a fix, the build process barely starts. I can try to create a pull request, but it is easier to just fix lines `81` and `101` in `beakerx/setup.py`

**Comment by scottdraves** [comment]
setup.py comes from https://github.com/jupyter-widgets/widget-cookiecutter/blob/master/%7B%7Bcookiecutter.github_project_name%7D%7D/setup.py
if it's wrong we should send a fix upstream.  would like to understand why/when it's needed though.
the cmd does not need any shell expansion or parsing.

**Comment by altavir** [comment]
I don't know, I am not using Python (I don't like it very much in fact). Just found the solution in the internet: [here](http://stackoverflow.com/questions/9531683/problems-using-subprocess-call-in-python-2-7-2-on-windows) for example.

EDIT: I believe there is an explanation down in answers [here](http://stackoverflow.com/questions/3022013/windows-cant-find-the-file-on-subprocess-call)

**Comment by altavir** [comment]
Another solution is to use either `subprocess.check_call(['cmd','/c','npm', '--version'])` or `subprocess.check_call(['powershell','-command','npm', '--version'])`

**Comment by gnestor** [comment]
@altavir Have you resolved this issue? I believe that including the `shell=True` argument is the correct way to deal with Windows compatibility. 

@scottdraves There is a relatively new package to abstract away Python packaging and enforce best practices: https://github.com/jupyter/jupyter-packaging

**Comment by scottdraves** [comment]
for the record the explanation from SO is:

On Windows, I believe the subprocess module doesn't look in the PATH unless you pass shell=True. However, shell=True can be a security risk if you're passing arguments that may come from outside your program. To make subprocess nonetheless able to find the correct executable, you can use shutil.which. Suppose the executable in your PATH is named frob:
```
subprocess.call([shutil.which('frob'), arg1, arg2])
```
(This works on Python 3.3 and above.)

**Comment by scottdraves** [comment]
thanks Grant we will have to look at that.  i don't have the bandwidth right now though.

**Comment by gnestor** [comment]
I'm happy to do the refactoring to needed to adopt it :+1:

**Comment by scottdraves** [comment]
ty altavir i am thinking your soln is best.

**Comment by altavir** [comment]
Even with this problem fixed, there is still an error later in the npm. I am not sure it is windows specific, but it prevents beakerx from completing installation.

**Comment by scottdraves** [comment]
we are going to take this stuff out of gradle soon anyway, so just the quick fix for now.

**Comment by altavir** [comment]
It is not working even with the fix, so I will be waiting for reliable installer.

**Comment by scottdraves** [comment]
that fix actually breaks it on unix.  it either has to be conditional on windows or something different.
also, that is not the only place where call_check is used on npm.

**Comment by konstantinTarletski** [comment]
While I tried to clone and install all project from zero (clean install).
I get errors.

I fix it that way:
My fix (commied PR created):

in file:
\beakerx\beakerx\build.gradle

I put this command first:
`project('js').tasks.install.execute()`

Was:
```
task install() {
  doLast {
    exec {
      commandLine(*(winOrUnix('pip', "install", "-e", ".")))
    }
    exec {
      commandLine(*(winOrUnix("jupyter", "nbextension", "install", "--py", "--symlink", "--sys-prefix", "beakerx")))
    }
    exec {
      commandLine(*(winOrUnix("jupyter", "nbextension", "enable", "--py", "--sys-prefix", "beakerx")))
    }
   project('js').tasks.install.execute()
  }
}
```

After fix:
```
task install() {
  doLast {
    project('js').tasks.install.execute()
    exec {
      commandLine(*(winOrUnix('pip', "install", "-e", ".")))
    }
    exec {
      commandLine(*(winOrUnix("jupyter", "nbextension", "install", "--py", "--symlink", "--sys-prefix", "beakerx")))
    }
    exec {
      commandLine(*(winOrUnix("jupyter", "nbextension", "enable", "--py", "--sys-prefix", "beakerx")))
    }
  }
}
```


Full step to install:
After I create fix I did "clean install" to try, is there any errors.
Do not get any errors.
How I install "beakerx":

```
git clone https://github.com/twosigma/beakerx.git
conda create -y -n beakerx python=3.5 jupyter pandas
activate beakerx
gradlew --no-daemon build
gradlew --no-daemon kernelInstall
gradlew --no-daemon :beakerx:install
```

Do not get any errors while this commands !

afrer that just :
`jupyter notebook` 

To start pplication.

**My confiduration:**

Windows 10 Pro
Version 1607
OS Build 14393.1066
System type 64-bit

(beakerx) C:\WorkspaceBeakerx\beakerx\demoFiles> java -version
java version "1.8.0_92"
Java(TM) SE Runtime Environment (build 1.8.0_92-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.92-b14, mixed mode)

(beakerx) C:\WorkspaceBeakerx\beakerx\demoFiles>npm -version
3.9.3

(beakerx) C:\WorkspaceBeakerx\beakerx\demoFiles>bower -v
1.7.9

(beakerx) C:\WorkspaceBeakerx\beakerx\demoFiles>conda list
 packages in environment at C:\Program Files\Anaconda3\envs\beakerx:

bleach                    1.5.0                    py35_0
colorama                  0.3.7                    py35_0
decorator                 4.0.11                   py35_0
entrypoints               0.2.2                    py35_1
html5lib                  0.999                    py35_0
icu                       57.1                     vc14_0  [vc14]
ipykernel                 4.6.1                    py35_0
ipython                   6.0.0                    py35_0
ipython_genutils          0.2.0                    py35_0
ipywidgets                6.0.0                    py35_0
jinja2                    2.9.6                    py35_0
jpeg                      9b                       vc14_0  [vc14]
jsonschema                2.6.0                    py35_0
jupyter                   1.0.0                    py35_3
jupyter_client            5.0.1                    py35_0
jupyter_console           5.1.0                    py35_0
jupyter_core              4.3.0                    py35_0
libpng                    1.6.27                   vc14_0  [vc14]
markupsafe                0.23                     py35_2
mistune                   0.7.4                    py35_0
mkl                       2017.0.1                      0
nbconvert                 5.1.1                    py35_0
nbformat                  4.3.0                    py35_0
notebook                  5.0.0                    py35_0
numpy                     1.12.1                   py35_0
openssl                   1.0.2k                   vc14_0  [vc14]
pandas                    0.19.2              np112py35_1
pandocfilters             1.4.1                    py35_0
path.py                   10.3.1                   py35_0
pickleshare               0.7.4                    py35_0
pip                       9.0.1                    py35_1
prompt_toolkit            1.0.14                   py35_0
pygments                  2.2.0                    py35_0
pyqt                      5.6.0                    py35_2
python                    3.5.3                         0
python-dateutil           2.6.0                    py35_0
pytz                      2017.2                   py35_0
pyzmq                     16.0.2                   py35_0
qt                        5.6.2                    vc14_3  [vc14]
qtconsole                 4.3.0                    py35_0
setuptools                27.2.0                   py35_1
simplegeneric             0.8.1                    py35_1
sip                       4.18                     py35_0
six                       1.10.0                   py35_0
testpath                  0.3                      py35_0
tornado                   4.5.1                    py35_0
traitlets                 4.3.2                    py35_0
vs2015_runtime            14.0.25123                    0
wcwidth                   0.1.7                    py35_0
wheel                     0.29.0                   py35_0
widgetsnbextension        2.0.0                    py35_0
win_unicode_console       0.5                      py35_0
zlib                      1.2.8                    vc14_3  [vc14]

**Comment by scottdraves** [comment]
this has been merged to master.  altavir, does it work for you?

**Comment by altavir** [comment]
No, still get:
```
npm ERR! node v4.4.4
npm ERR! npm  v2.15.1
npm ERR! code EPEERINVALID

npm ERR! peerinvalid The package webpack@1.15.0 does not satisfy its siblings' peerDependencies requirements!
npm ERR! peerinvalid Peer watch-ignore-webpack-plugin@1.0.0 wants webpack@1.5.x
npm ERR! peerinvalid Peer bower-webpack-plugin@0.1.9 wants webpack@>=1.4.0
npm ERR! peerinvalid Peer sass-loader@6.0.3 wants webpack@^2.0.0
```

I don't have time for clean install now. Will do it later.

By the way, `project('js').tasks.install.execute()` must be replaced by `dependsOn: ":beakerx:js:install"` in task declaration.
Also I do not understand the meaning of `--no-daemon` flags in commands. It changes nothing, just makes build process slower

**Comment by scottdraves** [comment]
the daemon can cause a problem if it's started with a different conda environment.

testing the build process requires a fresh clone, please LMK the results of this test.

thx.

**Comment by altavir** [comment]
Nope. Still not working even with clean install. The same error. Please fix #5074. Redownloading the repository takes a lot of time.

**Comment by scottdraves** [comment]
sorry to hear that.
you might want to try again after https://github.com/twosigma/beakerx/issues/5060

**Comment by altavir** [comment]
OK, it finally finishes installation, though there are still nasty warnings:
```
:beakerx:install
yarn install v0.24.5
[1/4] Resolving packages...
[2/4] Fetching packages...
warning fsevents@1.1.1: The platform "win32" is incompatible with this module.
info "fsevents@1.1.1" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
warning "sass-loader@6.0.5" has incorrect peer dependency "webpack@^2.0.0".
warning "watch-ignore-webpack-plugin@1.0.0" has incorrect peer dependency "webpack@1.5.x".
[4/4] Building fresh packages...
```
Examples still are not working, but at least it recognizes beaker classes. Maybe clean install will help.

**Comment by scottdraves** [comment]
thanks for the update.

**Comment by altavir** [comment]
I've got everything installed and running, but jupyter still does not display any beaker output (shows [*] and does nothing).
Maybe the problem is in this yarn warning:
```
[2/4] Fetching packages...
warning fsevents@1.1.1: The platform "win32" is incompatible with this module.
info "fsevents@1.1.1" is an optional dependency and failed compatibility check. Excluding it from installation.
```

**Comment by altavir** [comment]
I had to completely reinstalled the system due to hard disk failure. I reinstalled everything and finally it works!
Still, it does not work on other two other computers I tried and installation requires a lot of time and additional software (like node.js, yarn and clang).
I am going to use beakerx for lectures and interactive examples this autumn, I hope that working distribution will be available before that. Otherwise my students won't manage to install it.

I will close the issue for now.  It could be reopened if someone have similar problem.
